internal/dag/gatewayapi_processor.go

```diff
@@ -41,8 +42,42 @@ func (p *GatewayAPIProcessor) Run(dag *DAG, source *KubernetesCache) {
 		p.source = nil
 	}()
 
+	// Gateway must be defined for resources to be processed.
+	if p.source.gateway == nil {
+		p.Error("Gateway is not defined!")
+		return
+	}
+
 	for _, route := range p.source.httproutes {
-		p.computeHTTPRoute(route)
+
+		var validRoutes []*gatewayapi_v1alpha1.HTTPRoute
+
+		// Filter the HTTPRoutes that match the gateway which Contour is configured to watch.
+		// RouteBindingSelector defines a schema for associating routes with the Gateway.
+		// If Namespaces and Selector are defined, only routes matching both selectors are associated with the Gateway.
+
+		// ## RouteBindingSelector ##
+		//
+		// Selector specifies a set of route labels used for selecting routes to associate
+		// with the Gateway. If this Selector is defined, only routes matching the Selector
+		// are associated with the Gateway. An empty Selector matches all routes.
+		for _, listener := range p.source.gateway.Spec.Listeners {
+
+			if len(listener.Routes.Selector.MatchLabels) > 0 || len(listener.Routes.Selector.MatchExpressions) > 0 {
```

Comment:
Could this be moved into `match.LabelSelector`? Or should the empty case not generically match all resources?

Reply:
Seems like it already does? https://github.com/projectcontour/contour/pull/3415/files#diff-cc1654d51a4731d486b648439e46b022045d9bc596e7e5f0da32cf33e2ded3eaR26

```
if len(selector.MatchExpressions) == 0 {
	matchesExpression = true
} ...
```

 and the same for matching labels

Reply:
The Selector is not a pointer so we can't check against that. The API docs say that if it's not specified, then it matches all. 

I could refactor this out into a method to make it cleaner though.

Reply:
@sunjayBhatia I refactored this up a bit and I think reads much better.